### PR TITLE
Update all Python package versions by unpinning them.

### PR DESCRIPTION
**Pull request by deflaux**
Also:
* added package `plotnine` per #126
* replaced package `tensorflow` with `tensorflow_cpu` to get rid of the warnings about GPUs being unavailable for Terra Cloud Runtimes
* added package `google-resumable-media` as an explicit dependency to ensure a more recent version of it is used, pandas-gbq depends on it for table uploads
* `--use_rest_api` flag is now needed for `%%bigquery magic`
  * As of release [google-cloud-bigquery 1.26.0 (2020-07-20)](https://github.com/googleapis/python-bigquery/blob/master/CHANGELOG.md#1260-2020-07-20) the BigQuery Python client uses the BigQuery Storage client by default.
  * This currently causes an error on Terra Cloud Runtimes `the user does not have 'bigquery.readsessions.create' permission for '<Terra billing project id>'`.
  * To work-around this we uninstall the dependency `google-cloud-bigquery-storage` so that flag `--use_rest_api` can be used with `%%bigquery` to use the older, slower mechanism for data transfer.
* add `nbstripout` to `terra-jupyter-aou` and enable it globally
* improve test coverage by enabling tests that were intentionally commented out for the prior image

Smoke test pass for the images that were modified:
* [terra-jupyter-python](https://github.com/deflaux/terra-docker/actions/runs/625147084) smoke test results
* [terra-jupyter-aou](https://github.com/deflaux/terra-docker/actions/runs/625146524) smoke test results

The Hail and GATK images are unchanged by this PR, but just to demonstrate that this change will causes no future issues for them, in a test branch I bumped the version of `terra-jupyter-python` and those smoke tests pass too:
* [terra-jupyter-hail](https://github.com/deflaux/terra-docker/actions/runs/625153667) smoke test results
* [terra-jupyter-gatk](https://github.com/deflaux/terra-docker/actions/runs/625153665) smoke test results

### CODE REVIEW

**Comment by deflaux** [comment]
Also note that I wanted to make a more drastic change to `terra-jupyter-python/Dockerfile` similar to [this](https://github.com/deflaux/terra-docker/blob/bump-aou-versions/terra-jupyter-python/Dockerfile) but I resisted because (1) it would make the diff more difficult to read and (2) I wasn't sure if there was a particular reason for the individual `pip install` commands in the Dockerfile instead of one big `pip3 install --upgrade ` command.
```
RUN pip3 -V \
 && pip3 install --upgrade pip \
 && pip3 install --upgrade \
   biopython \
   bleach \
   bokeh \
   bx-python \
...
```

I could add a commit for that, if that cleanup is desired.

**Comment by deflaux** [comment]
> Also note that I wanted to make a more drastic change to `terra-jupyter-python/Dockerfile` similar to [this](https://github.com/deflaux/terra-docker/blob/bump-aou-versions/terra-jupyter-python/Dockerfile) but I resisted because (1) it would make the diff more difficult to read and (2) I wasn't sure if there was a particular reason for the individual `pip install` commands in the Dockerfile instead of one big `pip3 install --upgrade ` command.
> 
> ```
> RUN pip3 -V \
>  && pip3 install --upgrade pip \
>  && pip3 install --upgrade \
>    biopython \
>    bleach \
>    bokeh \
>    bx-python \
> ...
> ```
> 
> I could add a commit for that, if that cleanup is desired.

@calbach I am unable to add reviewers to PRs in this repo, but I wanted to call out that this changes the AoU image. In addition to using more recently Python package versions, it adds nbstripout and enables it globally for any git repo cloned to the machine.

File: terra-jupyter-aou/Dockerfile
Comment (calbach) [inline terra-jupyter-aou/Dockerfile:88]: You would also need to make updates to `CHANGELOG.md` and bump the version in `config.json`, see and example: https://github.com/DataBiosphere/terra-docker/pull/200/files

File: terra-jupyter-python/Dockerfile
Comment (calbach) [inline terra-jupyter-python/Dockerfile:33]: Per offline discussion I'd probably use a more explicit version lockfile of some kind (which can be easily regenerated) for the sake of determinism, but just my 2c - this doesn't directly affect me or the AoU image.

**Review by calbach — COMMENTED** [review]
Thanks for making the upgrades. The AoU changes LGTM, but Leo should review the base changes

File: terra-jupyter-aou/Dockerfile
Comment (deflaux) [reply terra-jupyter-aou/Dockerfile:88]: done!

File: terra-jupyter-python/Dockerfile
Comment (deflaux) [reply terra-jupyter-python/Dockerfile:33]: Thank you @calbach .  Leonardo team, what is your preference?

My goal with this PR is for the resulting images to have newer versions of all Python packages. Happy to change the details for how that is accomplished.

**Comment by Qi77Qi** [comment]
both `terra-jupyter-gatk` and `terra-jupyter-hail` depends on `terra-jupyter-python`, so you'll need to bump versions for those 2 as well

**Comment by deflaux** [comment]
> both `terra-jupyter-gatk` and `terra-jupyter-hail` depend on `terra-jupyter-python`, so you'll need to bump versions for those 2 as well

done!

> both `terra-jupyter-gatk` and `terra-jupyter-hail` depend on `terra-jupyter-python`, so you'll need to bump versions for those 2 as well

done!

**Comment by deflaux** [comment]
I added a few more tests for `gcloud`, `gsutil`, and `bq`. The newly added tests pass:
* https://github.com/deflaux/terra-docker/actions/runs/658460464
* https://github.com/deflaux/terra-docker/actions/runs/658412430
* https://github.com/deflaux/terra-docker/actions/runs/658348456
* https://github.com/deflaux/terra-docker/actions/runs/658411725

I think this PR is ready to merge unless any other changes are needed - thanks!

**Review by Qi77Qi — APPROVED** [review]
lgtm...but I'll leave this to @gpcarr since she's working on a related ticket

File: terra-jupyter-python/Dockerfile
Comment (rtitle) [reply terra-jupyter-python/Dockerfile:33]: I'm willing to try unpinning the versions

File: terra-jupyter-python/Dockerfile
Comment (rtitle) [inline terra-jupyter-python/Dockerfile:43]: Apparently this is fixed (I don't even remember this) -- I see there is a `pandas` test case 👍

**Comment by rtitle** [comment]
The Dockerfile and changelog updates look fine to me. 

I see the GitHub action tests failed on this PR -- is that because it's from a fork and there are auth issues (sorry if I missed some conversation). Can we ensure they pass before merge?

Also this PR is moderately risky -- we can have good test cases but still break somebody's code with a change like this. Merging this PR won't update anything in Terra/Leo -- but we should probably have a separate ticket to test on the Leo side (I expect we'll have to fix some automation tests) and write good release notes for our users.

**Comment by rtitle** [comment]
Thinking more, should this be a minor version update instead of a patch update?

https://semver.org/

That would also let us make patch fixes in case this is delayed by testing/user communication issues. I dunno, what do you think @Qi77Qi @gpcarr ?

File: terra-jupyter-aou/Dockerfile
Comment (gpcarr) [inline terra-jupyter-aou/Dockerfile:90]: Curious if we'd want what's in this directory to persist on the user's persistent disk? if so, everything under the directory `/home/$USER/notebooks` persists since that's where we mount the PD

File: terra-jupyter-python/Dockerfile
Comment (gpcarr) [reply terra-jupyter-python/Dockerfile:33]: Unpinning the versions sounds good to me too, we can do additional testing in Leonardo before releasing

**Comment by gpcarr** [comment]
> The Dockerfile and changelog updates look fine to me.
> 
> I see the GitHub action tests failed on this PR -- is that because it's from a fork and there are auth issues (sorry if I missed some conversation)? Can we ensure they pass before merge?
> 
> Also this PR is moderately risky -- we can have good test cases but still break somebody's code with a change like this. Merging this PR won't automatically update anything in Terra/Leo until we choose -- we should probably have a separate ticket to test this on the Leo side (I expect we'll have to fix some automation tests) and write good release notes for our users.

+1 on making a Leo ticket to test this and fix any tests that break and adding release notes. Also, we currently [pin](https://github.com/DataBiosphere/terra-docker/blob/master/terra-jupyter-base/Dockerfile#L118) the version of python in the base image so I wonder if updating all python packages might create some inconsistencies. We can keep an eye on this while testing

**Comment by gpcarr** [comment]
> Thinking more, should this be a minor version update instead of a patch update? (we've been reserving major version updates to language version changes like R 4.0.0)
> 
> https://semver.org/
> 
> That would also let us make patch fixes in case this is delayed by testing/user communication issues. I dunno, what do you think @Qi77Qi @gpcarr ?

I would +1 for a minor version bump since we aren't bumping the python version

**Comment by deflaux** [comment]
> Also note that I wanted to make a more drastic change to `terra-jupyter-python/Dockerfile` similar to [this](https://github.com/deflaux/terra-docker/blob/bump-aou-versions/terra-jupyter-python/Dockerfile) but I resisted because (1) it would make the diff more difficult to read and (2) I wasn't sure if there was a particular reason for the individual `pip install` commands in the Dockerfile instead of one big `pip3 install --upgrade ` command.
> 
> ```
> RUN pip3 -V \
>  && pip3 install --upgrade pip \
>  && pip3 install --upgrade \
>    biopython \
>    bleach \
>    bokeh \
>    bx-python \
> ...
> ```
> 
> I could add a commit for that, if that cleanup is desired.

@rtitile also let me know if a commit should be added to do the bigger cleanup of turning the many individual `pip install` commands into a single `pip3 install --upgrade pkg1 pkg2 pkg3 ...`. At the moment I'm not sure whether that change would yield a better, more self-consistent set of installed package versions since pip would get to compute the dependency tree all at once, but its definitely a nice cleanup.

**Comment by deflaux** [comment]
> > Thinking more, should this be a minor version update instead of a patch update? (we've been reserving major version updates to language version changes like R 4.0.0)
> > https://semver.org/
> > That would also let us make patch fixes in case this is delayed by testing/user communication issues. I dunno, what do you think @Qi77Qi @gpcarr ?
> 
> I would +1 for a minor version bump since we aren't bumping the python version

will do!

**Comment by deflaux** [comment]
> The Dockerfile and changelog updates look fine to me.
> 
> I see the GitHub action tests failed on this PR -- is that because it's from a fork and there are auth issues (sorry if I missed some conversation)? Can we ensure they pass before merge?
> 
> Also this PR is moderately risky -- we can have good test cases but still break somebody's code with a change like this. Merging this PR won't automatically update anything in Terra/Leo until we choose -- we should probably have a separate ticket to test this on the Leo side (I expect we'll have to fix some automation tests) and write good release notes for our users.

@rtitle do you want me to close this PR and resend from a branch instead of a fork? The tests will be able to function correctly if I do that and @Qi77Qi just gave me the needed permissions.

**Comment by Qi77Qi** [comment]
@deflaux I think that'd be great..just so we don't have to remember running the workflow after your PR is merged (not that we don't trust you)

**Comment by deflaux** [comment]
> > The Dockerfile and changelog updates look fine to me.
> > I see the GitHub action tests failed on this PR -- is that because it's from a fork and there are auth issues (sorry if I missed some conversation)? Can we ensure they pass before merge?
> > Also this PR is moderately risky -- we can have good test cases but still break somebody's code with a change like this. Merging this PR won't automatically update anything in Terra/Leo until we choose -- we should probably have a separate ticket to test this on the Leo side (I expect we'll have to fix some automation tests) and write good release notes for our users.
> 
> +1 on making a Leo ticket to test this and fix any tests that break and adding release notes. Also, we currently [pin](https://github.com/DataBiosphere/terra-docker/blob/master/terra-jupyter-base/Dockerfile#L118) the version of python in the base image so I wonder if updating all python packages might create some inconsistencies. We can keep an eye on this while testing

+1 it is a big change right now, but going forward it will make it a lot easier and less risky to provide fresh Python images with current package versions to users every quarter or so. The key thing, as you note, is to add new tests for anything we discover that was missed by the tests here in this GitHub action and your downstream tests.

Also consider applying the tests now in this git repo during your integration testing phase. The service account in GitHub Action environment does not have permissions identical to those of a Terra PET account, so some problems such as the lack of the `readsessions` permissions breaking the `%%biquery` Jupyter magic will not surface here.

File: terra-jupyter-aou/Dockerfile
Comment (deflaux) [reply terra-jupyter-aou/Dockerfile:90]: For AoU, we want this to persist on the base image so that its enabled globally by default for all users. It keeps people from accidentally checking into source control notebooks *containing outputs*.

**Comment by rtitle** [comment]
Re-submitting this PR from a DataBiosphere branch SGTM, would be good to get green tests.

I don't really have a strong opinion on
```
pip install pkg1 pkg2 pkg3
```
vs
```
pip install pkg1 && pip install pkg2 && pip install pkg3
```
I agree it would make the Dockerfile cleaner (and maybe more efficient to build) but don't think it would have a user-facing impact.

**Comment by deflaux** [comment]
Please see https://github.com/DataBiosphere/terra-docker/pull/207 for this code pushed to a branch of DataBiosphere/terra-docker. It's interesting to note that the checks are now all green on this PR from a fork. The commits are identical in both PRs, so I guess its de-duping the GitHub Action results.

**Comment by deflaux** [comment]
This PR was superseded by https://github.com/DataBiosphere/terra-docker/pull/207 which has been merged.

**Pull request closed without merging**
